Src/Extern/SUNDIALS/AMReX_SUNMemory.cpp

```diff
@@ -0,0 +1,140 @@
+#include <AMReX.H>
+#include <AMReX_Gpu.H>
+#include <AMReX_SUNMemory.H>
+#if defined(AMREX_USE_HIP)
+#include <sunmemory/sunmemory_hip.h>
+#elif defined(AMREX_USE_CUDA)
+#include <sunmemory/sunmemory_cuda.h>
+#elif defined(AMREX_USE_DPCPP)
+#include <sunmemory/sunmemory_sycl.h>
+#endif
+
+namespace amrex {
+namespace sundials {
+
+namespace {
+  int Alloc(SUNMemoryHelper, SUNMemory* memptr, size_t memsize, SUNMemoryType mem_type)
+  {
+    SUNMemory mem = SUNMemoryNewEmpty();
+
+    mem->ptr = NULL;
+    mem->own = SUNTRUE;
+    mem->type = mem_type;
+
+    if (mem_type == SUNMEMTYPE_HOST) {
+      mem->ptr = The_Cpu_Arena()->alloc(memsize);
+    } else if (mem_type == SUNMEMTYPE_UVM) {
+      mem->ptr = The_Managed_Arena()->alloc(memsize);
+    } else if (mem_type == SUNMEMTYPE_DEVICE) {
+      mem->ptr = The_Device_Arena()->alloc(memsize);
+    } else if (mem_type == SUNMEMTYPE_PINNED) {
+      mem->ptr = The_Pinned_Arena()->alloc(memsize);
+    } else {
+      free(mem);
+      return(-1);
+    }
+
+    *memptr = mem;
+    return(0);
+  }
+
+  int Dealloc(SUNMemoryHelper, SUNMemory mem)
+  {
+    if (mem == NULL) return(0);
+
+    if (mem->type == SUNMEMTYPE_HOST) {
+      if (mem->own) The_Cpu_Arena()->free(mem->ptr);
+    } else if (mem->type == SUNMEMTYPE_UVM) {
+    if (mem->own) The_Managed_Arena()->free(mem->ptr);
+    } else if (mem->type == SUNMEMTYPE_DEVICE) {
+      if (mem->own) The_Device_Arena()->free(mem->ptr);
+    } else if (mem->type == SUNMEMTYPE_PINNED) {
+      if (mem->own) The_Pinned_Arena()->free(mem->ptr);
+    } else {
+      return(-1);
```

Comment:
I assume this means that the caller is responsible for freeing `mem`.

Reply:
This should be

```c++
free(mem);
return -1;
```